internal/command/jsonformat/differ/list.go

```diff
@@ -0,0 +1,165 @@
+package differ
+
+import (
+	"reflect"
+
+	"github.com/zclconf/go-cty/cty"
+
+	"github.com/hashicorp/terraform/internal/command/jsonformat/change"
+	"github.com/hashicorp/terraform/internal/command/jsonprovider"
+)
+
+func (v Value) computeAttributeChangeAsList(elementType cty.Type) change.Change {
+	var elements []change.Change
+	current := v.getDefaultActionForIteration()
+	v.processList(elementType, func(value Value) {
+		element := value.ComputeChange(elementType)
+		elements = append(elements, element)
+		current = compareActions(current, element.GetAction())
+	})
+	return change.New(change.List(elements), current, v.replacePath())
+}
+
+func (v Value) computeAttributeChangeAsNestedList(attributes map[string]*jsonprovider.Attribute) change.Change {
+	var elements []change.Change
+	current := v.getDefaultActionForIteration()
+	v.processNestedList(func(value Value) {
+		element := value.ComputeChange(attributes)
+		elements = append(elements, element)
+		current = compareActions(current, element.GetAction())
+	})
+	return change.New(change.NestedList(elements), current, v.replacePath())
+}
+
+func (v Value) processNestedList(process func(value Value)) {
+	sliceValue := v.asSlice()
+	for ix := 0; ix < len(sliceValue.Before) || ix < len(sliceValue.After); ix++ {
+		process(sliceValue.getChild(ix, ix, false))
+	}
+}
+
+func (v Value) processList(elementType cty.Type, process func(value Value)) {
+	sliceValue := v.asSlice()
+
+	lcs := lcs(sliceValue.Before, sliceValue.After)
+	var beforeIx, afterIx, lcsIx int
+	for beforeIx < len(sliceValue.Before) || afterIx < len(sliceValue.After) || lcsIx < len(lcs) {
+		// Step through all the before values until we hit the next item in the
+		// longest common subsequence. We are going to just say that all of
+		// these have been deleted.
+		for beforeIx < len(sliceValue.Before) && (lcsIx >= len(lcs) || !reflect.DeepEqual(sliceValue.Before[beforeIx], lcs[lcsIx])) {
+			isObjectDiff := elementType.IsObjectType() && afterIx < len(sliceValue.After) && (lcsIx >= len(lcs) || !reflect.DeepEqual(sliceValue.After[afterIx], lcs[lcsIx]))
+			if isObjectDiff {
+				process(sliceValue.getChild(beforeIx, afterIx, false))
+				beforeIx++
+				afterIx++
+				continue
+			}
+
+			process(sliceValue.getChild(beforeIx, len(sliceValue.After), false))
+			beforeIx++
+		}
+
+		// Now, step through all the after values until hit the next item in the
+		// LCS. We are going to say that all of these have been created.
+		for afterIx < len(sliceValue.After) && (lcsIx >= len(lcs) || !reflect.DeepEqual(sliceValue.After[afterIx], lcs[lcsIx])) {
+			process(sliceValue.getChild(len(sliceValue.Before), afterIx, false))
+			afterIx++
+		}
+
+		// Finally, add the item in common as unchanged.
+		if lcsIx < len(lcs) {
+			process(sliceValue.getChild(beforeIx, afterIx, false))
+			beforeIx++
+			afterIx++
+			lcsIx++
+		}
+	}
+}
+
+func lcs(xs, ys []interface{}) []interface{} {
```

Comment:
This appears to be a separate implementation of the LCS algorithm we already have in this codebase (possibly based on the other diff renderer that I wrote, or maybe just a common ancestor). We might want to refer to the one in the command/format package for now, and ideally add more commentary around its semantics and the origin of the algorithm.

Reply:
Yes, there were two reasons I just copied across. The first was a concern over generics as the other one works over Value instead of interfaces, but I've asked about that and learned we can use generics. The second was a belief that the other one was going to be deleted when the old diff package goes away as I think that is the only place that references, so copying it over here was more of a long term move rather than a duplication. I've made a note of this, and will see if I can place it somewhere in common and get everywhere that references it to use it.